docs/developing.rst

```diff
@@ -0,0 +1,350 @@
+  * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either 
+  * express or implied. See the License for the specific language governing 
+  * permissions and limitations under the License.
+  */
+
+For example, here is the instructions for Intellij IDEA https://www.jetbrains.com/help/idea/copyright.html to add a Template and Scope profile using the template.
+
+Note that missing license header will be detected by Gradle license plugin and fails the build.
+
+
+Making Code Changes
+===================
+
+Project Strucure
+----------------
+
+The plugin codebase is in standard layout of Gradle project::
+
+ .
+ ├── CODE_OF_CONDUCT.md
+ ├── CONTRIBUTING.md
+ ├── LICENSE.TXT
+ ├── NOTICE
+ ├── README.md
+ ├── THIRD-PARTY
+ ├── build.gradle
+ ├── config
+ │   └── checkstyle
+ │       ├── checkstyle.xml
+ │       └── suppressions.xml
+ ├── docs
+ │   ├── dev
+ │   │   ├── SemanticAnalysis.md
+ │   │   ├── SubQuery.md
+ │   │   └── img
+ │   └── user
+ │       ├── admin
+ │       ├── index.rst
+ │       └── interfaces
+ ├── gradle.properties
+ ├── gradlew
+ ├── gradlew.bat
+ ├── opendistro-elasticsearch-sql.release-notes
+ ├── settings.gradle
+ └── src
+     ├── assembly
+     │   └── zip.xml
+     ├── main
+     │   ├── antlr
+     │   ├── java
+     │   └── resources
+     └── test
+         ├── java
+         └── resources
+
+Here are files and folders you are most likely to touch:
+
+- build.gradle: Gradle build script.
+- config/: only Checkstyle configuration files for now.
+- docs/: include documentation for developers and reference manual for users.
+- src/: source code root
+
+  - main/antlr: ANTLR4 grammar files.
+  - main/java: Java source code.
+  - test/java: Java test code.
+
+Code Convention
+---------------
+
+We’re integrated Checkstyle plugin into Gradle build: https://github.com/opendistro-for-elasticsearch/sql/blob/master/config/checkstyle/checkstyle.xml. So any violation will fail the build. You need to identify the offending code from Gradle error message and fix them and rerun the Gradle build. Here are the highlight of some Checkstyle rules:
+
+* No line starts with tab character in source file.
+* Line width <= 120 characters.
+* Wildcard imports: You can enforce single import by configuring your IDE. Instructions for Intellij IDEA: https://www.jetbrains.com/help/idea/creating-and-optimizing-imports.html#disable-wildcard-imports.
+* Operator needs to wrap at next line.
+
+
+Building and Running Tests
+==========================
+
+Gradle Build
+------------
+
+Most of the time you just need to run ./gradlew build which will make sure you pass all checks and testing. While you’re developing, you may want to run specific Gradle task only. In this case, you can run ./gradlew with task name which only triggers the task along with those it depends on. Here is a list for common tasks:
+
++---------------------------------+-----------------------------------------------------------+
```

Comment:
So you *can* do this, but yeesh is it hard to maintain. I recommend list tables: http://docutils.sourceforge.net/0.4/docs/ref/rst/directives.html#list-table